### ISSUE TITLE
$PATH gets appended twice

### ISSUE THREAD

**Issue by ykonoclast** [xonsh/xonsh#2537]
Hi, this one is weird : on Fedora26 when I open a terminal window (therefore, in graphical mode, from the desktop) and .xonshrc gets executed this line :
`$PATH.append('/home/ykonoclast/DevEnv/playground')`
Causes  $PATH to get _two_ '/home/ykonoclast/DevEnv/playground' entries at its end. 
This is only true for terminal _windows_ : in true tty mode the .xonshrc gets executed correctly resulting in no duplicate entries.

WHAT I TRIED :
* duplicate the "append" line in .xonshrc to see what happens : there are then 4 entries added to $PATH.
* insert an `input("press a key")` line in .xonshrc : as expected .xonshrc gets called only once when a terminal is opened, so the duplication seems not to be caused by double call.
* check aliases list : no duplicate here (but maybe there can't be, I'm not a pythonista, I don't know how the datastructure works)

**Comment by astronouth7303** [comment]
I can't reproduce this easily.

**Comment by ykonoclast** [comment]
Well.. can you reproduce it _hardly_?
Joke aside I tried deleting this line from xonshrc to check if another file gets called somewhere and already appends to $PATH : nada, its by groups of two or nothing (again, not in tty mode).

**Comment by scopatz** [comment]
Hi @ykonoclast - I wonder if your xonshrc is getting loaded twice for some reason. Can you put a `print("Loading...")` in your xonshrc and see how many times it gets printed?  Basically, I wonder if your system is launching xonsh and then your terminal emulator is launching xonsh inside of xonsh.

**Comment by ykonoclast** [comment]
Hi @scopatz and thanks for the advice. I tried it but, just like my first try (with an `input("press a key")`) I only see one output.

**Comment by scopatz** [comment]
Hmm Can you please post your .xonshrc? Or send it to me if you would like to keep it private?

**Comment by ykonoclast** [comment]
Here it is, nothing private (for the comments, pardon my French, literally...) :

```
#customisation du prompt
$PROMPT = '{env_name:{} }{BOLD_GREEN}{user}{BOLD_YELLOW}@{BOLD_RED}{hostname}{BOLD_BLUE} {cwd} {BOLD_YELLOW}{prompt_end}{NO_COLOR} '
$MULTILINE_PROMPT = '`·.,¸,.·*¯`·.,¸,.·*¯'

#alias
aliases['pdb'] = 'python3 -m pdb'
aliases['nano'] ='nano -AFLSilw'

#ajouts au path pour le mode interactif
$PATH.append('/home/ykonoclast/DevEnv/playground')

```

Hoping I'm not doing something ridiculous here...

**Comment by scopatz** [comment]
@ykonoclast - do you have xonsh set as your default shell? ie did you set it with chsh?

**Comment by ykonoclast** [comment]
Yes I did it following [these instructions](http://xon.sh/customization.html#set-xonsh-as-my-default-shell)

**Comment by astronouth7303** [comment]
Insert `print($PATH)` as the very first line of `.xonshrc` and see what that does.

**Comment by astronouth7303** [comment]
(I wonder if we should add a method to `$PATH` to handle uniqueness? Or possibly make it an ordered set?)

**Comment by ykonoclast** [comment]
> Insert print($PATH) as the very first line of .xonshrc and see what that does.

And we are progressing! _Ab initio_, the content of $PATH is 
`[<many usual things>,'/home/ykonoclast/DevEnv/playground']`
So indeed, when xonshrc is executed in a terminal emulator window, this value has already been appended before. But I don't know why this is not the case when I switch to a tty screen or why the line I tried at first (the `input`) doesn't stop execution before (I only see the `press a key..` part when I open a terminal window, certainly not at startup).

**Comment by astronouth7303** [comment]
A lot of things in the system use your login shell for various things, so my guess would be that something in the GUI login process starts something with your login shell. Especially if you're running v0.5.x, not v0.6.0.

**Comment by ykonoclast** [comment]
Indeed the package of my distro is 0.5.12

Shouldn't the `input` command stop execution of the rc file though? Or is it ignored in non-interactive environments? 

As it is, while the double entry is not really a problem, I guess I could simply test for the value and only append if it is not there.

**Comment by scopatz** [comment]
> Shouldn't the input command stop execution of the rc file though? Or is it ignored in non-interactive environments?

It is probably ignored in non-interactive environments.

I think the is probably room for adding some interface to for unique appends and inserts (though I don't think this should be the default behavior for plain-old-append() and insert()) . I do think that other tools have a strong expectation that PATHs are lists that they can add to and pop off of. So making it have hard uniqueness is probably not the most compatible way forward.

**Comment by scopatz** [comment]
@ykonoclast - would you mind please opening a new issue for unique appends, so that we can close this one out? Thanks!

**Comment by ykonoclast** [comment]
Ok [done here](https://github.com/xonsh/xonsh/issues/2562)

Should I close this issue myself or should it be done by someone else?

**Comment by scopatz** [comment]
Thanks!
